### ISSUE TITLE
BelongsTo can't save when parent resource use withCount in indexQuery

### ISSUE THREAD

**Issue by ziming** [laravel/nova-issues#2341]
- Laravel Version: 6.14
- Nova Version: 2.10.1
- PHP Version: 7.4.2
- Database Driver & Version: MySQL 8
- Operating System and Version: Mac OS Mojave
- Browser type and version: Chrome 80.0.3987.87

### Description:
In my UserResource I overrided indexQuery with something like this

```php
/**
     * Build an "index" query for the given resource.
     *
     * @param  \Laravel\Nova\Http\Requests\NovaRequest  $request
     * @param  \Illuminate\Database\Eloquent\Builder  $query
     * @return \Illuminate\Database\Eloquent\Builder
     */
    public static function indexQuery(NovaRequest $request, $query)
    {
        return $query->withCount('books');
    }
```

Then in my fields method I have something like this

```php
            Number::make('Books', function () {
                return $this->books_count;
            })->sortable(),
```

All well and good until I add a HasMany for Books

In the show page of a specific user. Add Book

Fill in all the entries. But when I save

The BelongsTo with book_id will get an error like 

This book may not be associated with this resource.

I investigated and see that this is a validation error thrown by the Relatable Rule.

I proceed to remove my indexQuery method and Number::make code and saving works again.

I think this is a bug as it should be able to save even if I use withCount in indexQuery

**Comment by cliffordjames** [comment]
Got exactly the same error when using withCount on the indexQuery with the parent model/resource.

**Comment by sebsel** [comment]
I have this too, but with a more specific error by Postgres (translated to the books example):

> SQLSTATE[HY093]: Invalid parameter number: parameter was not defined (SQL: select * from "books" where "books"."id" = 0 order by "books"."id" desc limit 1)

Turns out this is caused – for me – because one of the counted relationships has an extra parameter. Like so, translated to the books example:

```php
    public function books()
    {
        return $this->hasMany(Book::class)
            ->where('has_valid_isbn', true);
    }

```

... that `true` seems to be stripped out of the query by Nova, but is still in the bindings. Curious to know if that's the case for you too. (But probably not, because you state you use MySQL, and this only crashes in Postgres for me.)

I opened this issue on it: https://github.com/laravel/framework/issues/32526

The overwrite of the select-statement happens here:
https://github.com/laravel/nova/blob/ffc74319433ae817cb7c9edb1daa7f9144647d1c/src/Rules/Relatable.php#L49

**Comment by ziming** [comment]
ok thanks for the reminder stale. I shall bump it up then :)

**Comment by ziming** [comment]
Thank you @rbresjer I will try your temp fix soon

**Comment by crynobone** [comment]
Hello @ziming 

Wanted to get an updated state of the issue so I have a clearer idea how to move forward with this:

* Does the solution by @rbresjer solved the problem? 
* Wondering if having `Number::make('Books')` set to `exceptOnForms()` would have been better.
* `I investigated and see that this is a validation error thrown by the Relatable Rule.` Can you share what the exact validation error?

**Comment by ziming** [comment]
Yes, the code by @rbresjer solves the problem for index view. But not for detail view.

For index view I still can't sort the count though but being able to see the count is good enough for me.

I'm not sure about exceptOnForms, haven't tried that for this situation

For your last question I need to try again to see what is the validation error. Will try when I have the time.

**Comment by crynobone** [comment]
Tried reproducing the code on latest Laravel Nova v3 and unable to replicate the issue https://github.com/nova-issues/issue-2341/commit/a6e1754959665cbd9ceca69eb02682cee53d1b3a

It should be best to upgrade to latest version as we no longer releasing bugfix releases for v2.

**Comment by desaintflorent** [comment]
@crynobone I can confirm that the issue is still there in latest Nova version ( V3.20 ) for my case ( I'm changing default orberBy in the index query )
I can also confirm that rbresjer hack is working for me.

I had a quick look at your replicate code and here what you should edit in your indexQuery function to fire the bug   :

app/Nova/User.php 
```
public static $orderBy = ['posts_count' => 'desc'];
public static function indexQuery($request, $query)
{
          if (empty($request->get('orderBy'))) {
	          $query->getQuery()->orders = [];
	          $query->orderBy(key(static::$orderBy), reset(static::$orderBy));
          }
         $query->withCount('posts');
	 return $query;
}
```

**Comment by crynobone** [comment]
This has been fixed and released in v3.21.0
